draft-how-does-mutant-work.markdown

```diff
@@ -0,0 +1,1630 @@
+`matched_subjects` returns an array of `Enumerable<Subject>`, let's check out how this looks in our console:
+
+```
+[
+# The `subjects` array has 2 entries:
+# First subject:
+
+#<Mutant::Subject::Method::Instance 
+  context=
+    #<Mutant::Context 
+      scope=Greeter
+      source_path=#<Pathname:.../hello_world/lib/hello_world/greeter.rb>>
+  node=s(:def, :initialize,
+    s(:args,
+      s(:arg, :phrase)),
+        s(:ivasgn, :@phrase,
+          s(:lvar, :phrase)))
+
+>, # end of first subject
+
+# Second subject:
+
+#<Mutant::Subject::Method::Instance
+  context=
+    #<Mutant::Context
+      scope=Greeter
+      source_path=#<Pathname:.../hello_world/lib/hello_world/greeter.rb>>
+  node= s(:def, :say_hello,
+          s(:args,
+            s(:arg, :name)),
+          s(:if,
+            s(:ivar, :@enabled),
+            s(:dstr,          
+              # ...snip
+    > # end of second subject
+]
+```
+
+Just by reading through it you should see be able to get a feeling for what's happening, but let's go through it step by step:
+
+```
+Mutant::Subject::Method::Instance
+```
+
+The `Subject` class. As you can see it has 2 attributes:
+
+1.) A [context](https://github.com/mbj/mutant/blob/master/lib/mutant/context.rb)
+
+```
+  context=
+    #<Mutant::Context 
+      scope=Greeter
+      source_path=#<Pathname:.../hello_world/lib/hello_world/greeter.rb>>
+```
+
+which contains:
+  * a [scope](https://github.com/mbj/mutant/blob/master/lib/mutant/scope.rb) called `Greeter`. This should look familiar to you now, doesn't it? That's the class from above we're actually testing.
+  * a `source_path`: Again, no magic here, you can see the path to the "greeter.rb" on my local file system
+
+2.) A [node](https://github.com/mbj/mutant/blob/master/lib/mutant/mutator/node.rb)
+
+```
+  node=s(:def, :say_hello,
+         s(:args,
+          # ...snip
+```
+
+representing an [abstract syntax trees (AST)](https://en.wikipedia.org/wiki/Abstract_syntax_tree) consisting of [S-expressions](https://en.wikipedia.org/wiki/S-expression).
+
+But even without knowing **anything** about ASTs or S-expressions by looking at this
+
+```
+  node=s(:def, :initialize,
+```
+
+or this
+
+```
+node=s(:def, :say_hello,
+```
+
+you can kind of figure out that both nodes correspond to the methods in our `Greeter` class:
+
+```Ruby
+class Greeter
+  def initialize(phrase)
+    @enabled = true
+    @phrase = phrase
+  end
+
+  def say_hello(name)
+    "#{@phrase} #{name}" if @enabled
+  end
+end
+```
+
+This
+
+```
+s(:def, :say_hello,
+  s(:args,
+    # ...snip
+```
+
+is the abstract syntax tree for this code in form of S-expressions.
+
+__Alright, so this doesn't look so complex anymore, now does it?__
+
+* We have a class with 2 instance methods
+* `Mutant` takes this class and turns it into 2 subjects
+* Where each subject contains a node consisting of S-Expressions that corresponds to one of our instance methods.
```

Comment:
OT: The details on how mutant figures out the source location and the actual AST nodes that represent the method are also interesting, but out of scope here.